apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/GoogleLogger.cs

```diff
@@ -0,0 +1,310 @@
+﻿// Copyright 2021 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.Logging.V2;
+using Google.Protobuf.WellKnownTypes;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using static System.FormattableString;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// <see cref="ILogger"/> for Google Cloud Logging.
+    /// </summary>
+    public sealed class GoogleLogger : ILogger
+    {
+        private const string GcpConsoleLogsBaseUrl = "https://console.cloud.google.com/logs/viewer";
+
+        /// <summary>The log name given when creating the logger.</summary>
+        private readonly string _logName;
+
+        /// <summary>The consumer to push logs to.</summary>
+        private readonly IConsumer<LogEntry> _consumer;
+
+        /// <summary>The trace target or null if non exists.</summary>
+        private readonly TraceTarget _traceTarget;
+
+        /// <summary>
+        /// The log target, indicating mainly if the target is a project or an organization.
+        /// </summary>
+        private readonly LogTarget _logTarget;
+
+        /// <summary>The logger options.</summary>
+        private readonly LoggerOptions _loggerOptions;
+
+        /// <summary>The formatted log name.</summary>
+        private readonly string _fullLogName;
+
+        /// <summary>A clock for getting the current timestamp.</summary>
+        private readonly IClock _clock;
+
+        /// <summary>The service provider to resolve additional services from.</summary>
+        private readonly IServiceProvider _serviceProvider;
+
+        private readonly Action<IServiceProvider, LogEntry, TraceTarget> _obsoleteTraceContextGetter;
+
+        private readonly AmbientScopeManager _ambientScopeManager;
+
+        internal GoogleLogger(
+            IConsumer<LogEntry> consumer, LogTarget logTarget, LoggerOptions loggerOptions,
+            string logName, IClock clock = null, IServiceProvider serviceProvider = null)
+#pragma warning disable CS0618 // Type or member is obsolete
+            : this (consumer, logTarget, loggerOptions, logName, null, null, clock, serviceProvider)
+#pragma warning restore CS0618 // Type or member is obsolete
+        {
+        }
+
+        [Obsolete("Added for backward compatibility only when moving GoogleLogger to Common.")]
```

Comment:
I'm somewhat lost as to whether we actually still need this. Given that it's internal, I'd expect to be able to remove it?

Reply:
This is the one remaining piece that allows us to use the Common.GoogleLogger from the AspNetCore.GoogleLogger taking into account some of the obsolte types that remained in AspNetCore* that were not moved to Common. This is called from Common.GoogleLoggleProvider which is called from AspNetCore.GoogleLoggerProvider. I added the Obsolete attribute so that we ourselves know not to depend on it.